syntax = "proto3";
import "google/protobuf/timestamp.proto";
import "google/protobuf/any.proto";

package mlayerregister;

/*
   Protobuf definition for M Layer Register Specification v0.03
   
   M-layer expression, q, defined as: 
   q = <q>{q}[[q]]
   <q> is the aspect (generalized kind of quantity)
   {q} is the value (numerical value or name of nominal type)
   [[q]] is the scale, which includes the scale-type and reference (unit)

   M-layer objects
   Aspect
    ID
    Name
    Symbol
    Definition

   Scale
    ID
    Name
    Unit or reference
    Scale type

   Unit
    ID
    Name
    Symbol
    Definition
  
  Scale type
    Ratio
    Interval
    BoundedInterval
    Ordinal
    Nominal

  AspectScale
    AspectID
    ScaleID

  Register Model - implement as Digital Objects
  Allow M-Layer objects to be abstracted as Digital Objects
  Identifiers held in DO identifier
  DigitalObject
    ID
    type
    metadata
    content (m-layer object)

  M Layer Expression
  Value, Digital Object (AspectScale) 
  Value, Digital Object (Scale)
*/

enum ScaleType {
  RATIO = 0;
  INTERVAL = 1;
  ORDINAL = 2;
}

message Register {
  string name = 1;
  string uuid = 2;
  repeated AspectObject aspects = 3;
  repeated ScaleObject scales = 4;
  repeated MObject mobjects = 5;
  repeated ConversionObject converters = 6;
}

/* 
   See Cordra DO implementation
   gitlab.com/cnri/cordra/.../cordra-core/src/main/java/net/cnri/cordra/api/CordraObject.java

   Rather than JsonElement use protobuf.Any
   JsonElement content - assume this is the specific m-layer object
   
   not included
   JsonObject userMetadata
   Payloads (part of the Digital Object specification)
   */

message DigitalObject {
  string id = 1;
  string type = 2;
  Metadata metadata = 3;
  google.protobuf.Any content = 4;
}

/* 
   Cordra Object Metadata
   */

message Metadata {
  int64 createdOn = 1;
  string createdBy = 2;
  int64 modifiedOn = 3;
  string modifiedBy = 4;
  bool isVersion = 5;
  string versionOf = 6;
  string publishedBy = 7;

}

/*
Cordra Payload

class Payload
string name
string filename
string mediaType
long size
java.io.InputStream

   */

/*
PID Kernel
Kernel information for a persistent identifier
www.rd-allliance.org/system/files/RDA Recommendation on PID Kernel Information_final.pdf

PID                       Handle
KernelInformationProfile  Handle
digitalObjectType         Handle
digitalObjectLocation     URL
digitalObjectPolicy       Handle
eTag                      Hex string
dataModified              ISO 8601
dateCreated
version                   string
wasDerivedFrom            PROV-DM properties

*/

/*
   PID Type 

*/

/* ObjectStore
   */

message DigitalObjectStore {
  string name = 1;
  string uuid = 2;
  string parent_uuid = 3;
  string address = 4;
  DigitalObjectStoreInfo info = 5;
}

message DigitalObjectStoreInfo {
  google.protobuf.Timestamp created = 1;
  DigitalObjectStoreAuxInfo aux = 2;
  repeated DigitalObject objects = 3;
}

message DigitalObjectStoreAuxInfo {
  string description = 1;
}

message Identifier {
  string name = 1;
  string value = 2;
  string type = 3;
  string description = 4;
}

message AlternateIdentifier {
  string value = 1;
  string type = 2;
  string description = 3;
}

message Context {
  string locale = 1;
  string name = 2;
  string symbol = 3;
}

message MLayerMsg {
  string identifier = 1;
  string uuid = 2;
  string name = 3;
  float value = 4;
  MObject mobject = 5;
  ScaleObject scale = 6;
}

message AspectObject {
  string identifier = 1;
  string uuid = 2;
  string name = 3;
  string symbol = 4;
  string definition = 5;
  string locale = 6;
}

message ScaleObject {
  string identifier = 1;
  string uuid = 2;
  string name = 3;
  string symbol = 4;
  ScaleType scaletype = 5;
  string referenceid = 6;
  string referenceuuid = 7;
}

message AspectScaleObject {
  string aspectid = 1;
  string scaleid = 2;
}

/*
   Reference or Unit, for example SI unit for length meter, m
   */

message ReferenceObject {
  string identifier = 1;
  string uuid = 2;
  string name = 3;
  string symbol = 4;
  string definition = 5;
  string locale = 6;
  string system_id = 7;
  string system_uuid = 8;
  repeated int32 dimensions = 9;
  int32 prefix = 10;
  string alternateid = 11;
  string alternateid_type = 12;
  string alternateid_description = 13;
}


message MObject {
  string identifier = 1;
  string uuid = 2;
  string name = 3;
  ScaleObject scale = 4;
  AspectObject aspect = 5;
}

message ConversionObject {
  string identifier = 1;
  string uuid = 2;
  string name = 3;
  string srcid = 4;
  string dstid = 5;
  ScaleType scaletype = 6;
  int32 nparms = 7;
  repeated float parameters = 8;
}

/*
   Updated schemas
*/

message AspectObjectv2 {
  Identifier id = 1;
  Context context = 2;
  string definition = 3;
}

message ScaleObjectv2 {
  Identifier id = 1;
  Context context = 2;
  ScaleType scaletype = 5;
  Identifier reference_id = 6;
}

message ReferenceObjectv2 {
  Identifier id = 1;
  Context context = 2;
  string definition = 3;
  Identifier system_id = 4;
  Identifier alternate_id = 5;
  repeated int32 dimensions = 6;
  int32 prefix = 7;

}

message UnitSystemObject {
  Identifier id = 1;
  string name = 2;
  repeated Identifier basis = 3;
}

/* Conversion between two scales of the same type 
   with the same aspect
 */
message ConversionObjectv2 {
  Identifier id = 1;
  Identifier src_id = 2; /* ScaleObject identifier */
  Identifier dst_id = 3; /* ScaleObject identifier */
  repeated float factors = 5;
}

/* Conversion between two scales of different type 
   with the same aspect
 */

message ScalesForAspectsv2 {
  Identifier id = 1;
  Identifier aspect_id = 2;
  Identifier src_id = 3;
  Identifier dst_id = 4;
  repeated float factors = 6;
}

/* M-layer scales should be expressed as AspectScales
   convenience to not express the Aspect can be handled
   by ScaleForAspect
   */
message AspectScaleObjectv2 {
  Identifier id = 1;
  Identifier aspect_id = 2;
  Identifier scale_id = 3;
}

/* Support to express data as a different ScaleAspect 
   Aspect may change
   Scale may change
   Requires expressing data with AspectScales
   */

message CastingObjectv2 {
  Identifier id = 1;
  Identifier srd_id = 2;
  Identifier dst_id = 3;
  string function = 4;
  repeated float parameters = 5;
  
}
